### PR TITLE
planner/cascades: add transformation PushSelDownJoin & Implementation for HashJoin

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 
1. add Transformation rule `PushSelDownJoin`, but currently only consider innerJoin.
2. add ImplementationRule for HashJoin for innerJoin.

#### Transformation rule: PushSelDownJoin
This PR ports some codes from `rule_predicate_push_down` to cascades framework, but only handles innerJoin. 
#### Implementation rule: ImplHashJoinBuildLeft & implHashJoinBuildRight
This PR also adds two Implementation rules for Join(only innerJoin).
#### Cascades Framework
Because the transformation rule may return a TableDual, this PR implements the logic to handle `eraseAll` returned from the rule by adding a new method `DeleteAll` for `Group`. 
### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -451,3 +454,88 @@ func (r *TransformLimitToTopN) OnTransform(old *memo.ExprIter) (newExprs []*memo
...
+		join.EqualConditions = equalCond
+		join.OtherConditions = otherCond
+		leftCond = leftPushCond
+		rightCond = rightPushCond
+	default:
+		// TODO: Enhance this rule to deal with LeftOuter/RightOuter/Semi/SmiAnti/LeftOuterSemi/LeftOuterSemiAnti Joins.
+	}
+	leftCond = expression.RemoveDupExprs(sctx, leftCond)
+	rightCond = expression.RemoveDupExprs(sctx, rightCond)
+	for _, eqCond := range join.EqualConditions {
```
Comment (lamxTyler) [inline planner/cascades/transformation_rules.go]: Why don't we need to `updateEQCond` like the orignal code do?

File: planner/cascades/implementation_rules.go [hunk 2]
```diff
@@ -334,3 +338,66 @@ func (r *ImplTopNAsLimit) OnImplement(expr *memo.GroupExpr, reqProp *property.Ph
 	}.Init(lt.SCtx(), expr.Group.Prop.Stats, lt.SelectBlockOffset(), newProp)
 	return impl.NewLimitImpl(physicalLimit), nil
 }
+
+func getImplForHashJoin(expr *memo.GroupExpr, prop *property.PhysicalProperty, innerIdx int, useOuterToBuild bool) memo.Implementation {
```
[the remark this answers is not in the document]
Comment (francis0407) [reply planner/cascades/implementation_rules.go]: For innerJoin, we only need to use the inner side to build the hash table. `useOuterToBuild` only used for outerJoins.

File: planner/cascades/transformation_rules.go (hunk 1 above)
Comment (francis0407) [reply planner/cascades/transformation_rules.go]: `updateEQCond` adds extra `Projection` below if there are `Equal` expressions whose args are not `Column`. It requires a big change to be implemented here. I'm gonna add a TODO here.

File: planner/cascades/implementation_rules.go [hunk 3]
```diff
@@ -64,6 +64,10 @@ var defaultImplementationMap = map[memo.Operand][]ImplementationRule{
 		&ImplTopN{},
 		&ImplTopNAsLimit{},
 	},
+	memo.OperandJoin: {
+		&ImplHashJoinBuildLeft{},
+		&ImplHashJoinBuildRight{},
```
Comment (lzmhhh123) [inline planner/cascades/implementation_rules.go]: Are the `outer` and `inner` better than the names `left` and `right`?

Comment (francis0407) [reply planner/cascades/implementation_rules.go]: For other join types, `outer` and `inner` is better. But I think it doesn't make sense for `InnerJoin`. While, `left` and `right` have meaning for all of them.

**Pull request merged**